apps/src/blockly/addons/cdoFieldButton.js

```diff
@@ -68,4 +69,14 @@ export default class CdoFieldButton extends GoogleBlockly.Field {
       this.buttonElement_.setAttribute('width', this.size_.width + 8);
     }
   }
+
+  positionBorderRect_() {
+    super.positionBorderRect_();
+    if (this.buttonElement_) {
```

Comment:
I think you want to do some of this work in an `applyColour` [method](https://developers.google.com/blockly/guides/create-custom-blocks/fields/customizing-fields/creating#matching_block_colours). The block will automatically call this method when the block is rendering and it's where internally we do some related kinds of rendering updates. Specifically if you need to update the colors of a field based on its parent block, that's what this method is for. So the last two lines can go in `applyColour`.


As for setting the height and width, that's where I'm a little confused. You have this `updateWidth_` method defined, and it says it calls `super`, but `GoogleBlockly.Field` doesn't have a method called `updateWidth_`. The cdo fork does, but I don't see how that could be relevant here since we inherit directly from `GoogleBlockly.Field`. My guess is this `updateWidth_` method is totally unused and if you tried to call it, you would get an error because `super.updateWidth_()` does not exist.

Also, in that function, you set the width differently than how you're setting it in this new code, and that's confusing to me, but maybe the calculation is just out of date since the `updateWidth_` calculation is not used anywhere.

Instead, to update the size of the field, you can do that work in `updateSize_`. You will still need to call `super.updateSize_` as well. Then you can set the width and height attributes in that method.

TLDR: Update the height and width properties in `updateSize_` and update the fill colors in `applyColours` (note the uk spelling of colour)

Reply:
For more info on why: 

The current code works because `updateSize_` (which is called when the field value changes or when the block is re-rendered e.g. when the renderer/theme changes) calls `positionBorderRect`. So you've found a method that is called at the appropriate time and tacked on some unrelated work to it, but it would be better to put that work in a more appropriate spot.


`applyColour` is called only when the color of the field might need to change, e.g. parent block changed, `changeColour` was called on a block directly, or the renderer changed. That is a better time to update the fill colors because you don't need to update them every time the value of the field changes like you would be doing with the current code.

Reply:
This feels much better, thank you Maribeth! I also confirmed calling `super.updateWidth_()` breaks, which makes sense -- fully removed that function.

Curious where I should have looked to see the `applyColour` info in the docs (seems quite obvious based on the link you shared 🤦 ) -- I didn't see any reference to it in the `Field` docs?

Reply:
How to find stuff in the docs: I always look in the Guides section first, and there are several pages about fields. The bulk of the info about the APIs are on the [Creating a Custom Field](https://developers.google.com/blockly/guides/create-custom-blocks/fields/customizing-fields/creating) page.

If I don't see anything in the Guides section, I'll check Reference then, but this is a bit more spotty as it's auto-generated. I noticed the `applyColour` method is not there so I'll fix that for next release.

Reply:
Cool I always go Reference first, good to know I should start with Guides!